rplugin/python3/defx/base/kind.py

```diff
@@ -18,18 +20,21 @@
 from defx.session import Session
 from defx.view import View
 
-_action_table: typing.Dict[str, ActionTable] = {}
-
-ACTION_FUNC = typing.Callable[[View, Defx, Context], None]
+Kind = typing.Any
+ACTION_FUNC = typing.Callable[[Kind, View, Defx, Context], None]
 
 
 def action(name: str, attr: ActionAttr = ActionAttr.NONE
            ) -> typing.Callable[[ACTION_FUNC], ACTION_FUNC]:
     def wrapper(func: ACTION_FUNC) -> ACTION_FUNC:
-        _action_table[name] = ActionTable(func=func, attr=attr)
-
-        def inner_wrapper(view: View, defx: Defx, context: Context) -> None:
-            return func(view, defx, context)
+        func._is_action = True  # type:ignore
```

Comment:
Why the type is ignored?

Reply:
Because setting attribute to function throws mypy error.
> "Callable[[Any, View, Defx, Context], None]" has no attribute "_is_action"

I learned this code from pynvim's decorator, but I don't know whether this is the best way.

Reply:
OK.  So the type definition is wrong.
It is not just callable.

Reply:
This is the better.

```diff
diff --git a/rplugin/python3/defx/base/kind.py b/rplugin/python3/defx/base/kind.py
index 389c1dd..7406256 100644
--- a/rplugin/python3/defx/base/kind.py
+++ b/rplugin/python3/defx/base/kind.py
@@ -23,13 +23,26 @@ _action_table: typing.Dict[str, ActionTable] = {}
 ACTION_FUNC = typing.Callable[[View, Defx, Context], None]


+class ActionFunc:
+    def __init__(self, name: str, attr: ActionAttr, func: ACTION_FUNC):
+        self._is_action = True
+        self._name = name
+        self._attr = attr
+        self._func = func
+
+    def __call__(self, view: View, defx: Defx, context: Context) -> None:
+        return self._func(view, defx, context)
+
+
 def action(name: str, attr: ActionAttr = ActionAttr.NONE
            ) -> typing.Callable[[ACTION_FUNC], ACTION_FUNC]:
     def wrapper(func: ACTION_FUNC) -> ACTION_FUNC:
+        f = ActionFunc(name, attr, func)
+
         _action_table[name] = ActionTable(func=func, attr=attr)

         def inner_wrapper(view: View, defx: Defx, context: Context) -> None:
-            return func(view, defx, context)
+            return f(view, defx, context)
         return inner_wrapper
     return wrapper
```

Reply:
Thanks! Fixed in https://github.com/Shougo/defx.nvim/pull/299/commits/14528da30209a18546def0e64dbb38da03ca97d7 .